### PR TITLE
Added Curator and Spark example

### PR DESCRIPTION
**Pull request by ronjer30**
## Description
Adds a simple example to demonstrate how Apache Spark can interoperate with NeMo Curator modules using intermediate parquet files.  

## Usage
Jupyter notebook tutorial

## Checklist

- [X] I am familiar with the [Contributing Guide](https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md).
- [ ] New or Existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: experiments/curator_apache_spark/curator_and_pyspark_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,963 @@
...
+   "cell_type": "code",
+   "execution_count": 61,
+   "id": "d9c348ec",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "spark = SparkSession.builder \\\n",
+    "    .master(\"local[4]\") \\\n",
+    "    .appName(\"Data Curation\") \\\n",
+    "    .config(\"spark.driver.memory\", '4g') \\\n",
```
Comment (sarahyurick) [inline experiments/curator_apache_spark/curator_and_pyspark_example.ipynb:670]:
```suggestion
    "    .config(\"spark.driver.memory\", "4g") \\\n",
```

File: experiments/curator_apache_spark/curator_and_pyspark_example.ipynb [hunk 2]
```diff
@@ -0,0 +1,963 @@
...
+    "    .config(\"spark.executor.memory\", '4g') \\\n",
```
Comment (sarahyurick) [inline experiments/curator_apache_spark/curator_and_pyspark_example.ipynb:671]:
```suggestion
    "    .config(\"spark.executor.memory\", "4g") \\\n",
```

File: experiments/curator_apache_spark/curator_and_pyspark_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,963 @@
...
+      "+--------------------+--------------------+--------------------+---------+--------------+\n",
+      "only showing top 5 rows\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "from pyspark.sql.functions import size, split, length, expr\n",
+    "# Calculate Word Count\n",
+    "df = df.withColumn(\"WordCount\", size(split(df[\"text\"], r'\\s+')))\n",
```
Comment (sarahyurick) [inline experiments/curator_apache_spark/curator_and_pyspark_example.ipynb:748]:
```suggestion
    "df = df.withColumn(\"WordCount\", size(split(df[\"text\"], r"\\s+")))\n",
```

File: experiments/curator_apache_spark/curator_and_pyspark_example.ipynb [hunk 4]
```diff
@@ -0,0 +1,963 @@
...
+   "source": [
+    "LOG_DIR = os.path.join(DATA_DIR, \"logs\")\n",
+    "EXACT_DEDUP_OUT_DIR = os.path.join(DATA_DIR, \"exact_dedup\")\n",
+    "\n",
+    "!mkdir -p {LOG_DIR}\n",
+    "!mkdir -p {EXACT_DEDUP_OUT_DIR}\n",
+    "\n",
+    "#ignores checksum and marker files created by Spark job\n",
+    "processed_files = [filename for filename in get_all_files_paths_under(PROCESSED_DIR) \n",
+    "                   if not filename.endswith('.crc') or filename.endswith('_SUCCESS')]\n"
```
Comment (sarahyurick) [inline experiments/curator_apache_spark/curator_and_pyspark_example.ipynb:849]:
```suggestion
    "                   if not filename.endswith(".crc") or filename.endswith("_SUCCESS")]\n"
```

File: experiments/curator_apache_spark/curator_and_pyspark_example.ipynb [hunk 5]
```diff
@@ -0,0 +1,963 @@
...
+   "execution_count": null,
+   "id": "dfaaa765",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "t0 = time.time()\n",
+    "# Read input dataset from Spark output\n",
+    "input_dataset = DocumentDataset.read_parquet(processed_files, backend='cudf')\n",
```
Comment (sarahyurick) [inline experiments/curator_apache_spark/curator_and_pyspark_example.ipynb:863]:
```suggestion
    "input_dataset = DocumentDataset.read_parquet(processed_files, backend="cudf")\n",
```

File: experiments/curator_apache_spark/curator_and_pyspark_example.ipynb [hunk 6]
```diff
@@ -0,0 +1,963 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "LOG_DIR = os.path.join(DATA_DIR, \"logs\")\n",
+    "EXACT_DEDUP_OUT_DIR = os.path.join(DATA_DIR, \"exact_dedup\")\n",
+    "\n",
+    "!mkdir -p {LOG_DIR}\n",
+    "!mkdir -p {EXACT_DEDUP_OUT_DIR}\n",
+    "\n",
+    "#ignores checksum and marker files created by Spark job\n",
```
Comment (sarahyurick) [inline experiments/curator_apache_spark/curator_and_pyspark_example.ipynb:847]:
```suggestion
    "# Ignores checksum and marker files created by Spark job\n",
```

File: experiments/curator_apache_spark/curator_and_pyspark_example.ipynb [hunk 7]
```diff
@@ -0,0 +1,963 @@
...
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "t0 = time.time()\n",
+    "# Read input dataset from Spark output\n",
+    "input_dataset = DocumentDataset.read_parquet(processed_files, backend='cudf')\n",
+    "\n",
+    "#Run exact deduplication to the input\n",
```
Comment (sarahyurick) [inline experiments/curator_apache_spark/curator_and_pyspark_example.ipynb:865]:
```suggestion
    "# Run exact deduplication on input_dataset\n",
```

File: experiments/curator_apache_spark/curator_and_pyspark_example.ipynb [hunk 8]
```diff
@@ -0,0 +1,963 @@
...
+    "exact_dup = ExactDuplicates(\n",
+    "    logger=LOG_DIR,\n",
+    "    id_field=\"id\",\n",
+    "    text_field=\"text\",\n",
+    "    hash_method=\"md5\",\n",
+    "    cache_dir=EXACT_DEDUP_OUT_DIR #Duplicated document ID list is output to the cache_dir\n",
```
Comment (sarahyurick) [inline experiments/curator_apache_spark/curator_and_pyspark_example.ipynb:871]:
```suggestion
    "    cache_dir=EXACT_DEDUP_OUT_DIR # Duplicated document ID list is outputted to the cache_dir\n",
```

**Comment by ronjer30** [comment]
As discussed, moving this example to docs with a new PR [https://github.com/NVIDIA/NeMo-Curator/pull/261](url)

**Pull request closed without merging**